crates/cext/src/dag.rs

```diff
@@ -151,12 +157,553 @@ pub unsafe extern "C" fn qk_dag_num_qubits(dag: *const DAGCircuit) -> u32 {
+///
+/// @param dag A pointer to the DAG to apply the gate to.
+/// @param gate The StandardGate to apply.
+/// @param qubits The pointer to the array of ``uint32_t`` qubit indices to add the gate on. This
+///     can be a null pointer if there are no qubits for ``gate`` (e.g. ``QkGate_GlobalPhase``).
+/// @param params The pointer to the array of ``double`` values to use for the gate parameters.
+///     This can be a null pointer if there are no parameters for ``gate`` (e.g. ``QkGate_H``).
+/// @param front If ``true``, the gate is applied as the first operation on the specified qubits,
+///     rather than as the last.
+///
+/// @return The index of the newly added operation node.
+///
+/// # Example
+/// ```c
+/// QkDag *dag = qk_dag_new();
+/// QkQuantumRegister *qr = qk_quantum_register_new(1, "my_register");
+/// qk_dag_add_quantum_register(dag, qr);
+///
+/// uint32_t qubit[1] = {0};
+/// qk_dag_apply_gate(dag, QkGate_H, qubit, NULL, false);
+///
+/// qk_dag_free(dag);
+/// qk_quantum_register_free(qr);
+/// ```
+///
+/// # Safety
+///
+/// The ``qubits`` and ``params`` types are expected to be a pointer to an array of ``uint32_t``
+/// and ``double`` respectively where the length is matching the expectations for the standard
+/// gate. If the array is insufficiently long the behavior of this function is undefined as this
+/// will read outside the bounds of the array. It can be a null pointer if there are no qubits
+/// or params for a given gate. You can check ``qk_gate_num_qubits`` and ``qk_gate_num_params`` to
+/// determine how many qubits and params are required for a given gate.
+///
+/// Behavior is undefined if ``dag`` is not a valid, non-null pointer to a ``QkDag``.
+#[unsafe(no_mangle)]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_dag_apply_gate(
+    dag: *mut DAGCircuit,
+    gate: StandardGate,
+    qubits: *const u32,
+    params: *const f64,
+    front: bool,
+) -> u32 {
+    // SAFETY: Per documentation, the pointer is to valid data.
+    let dag = unsafe { mut_ptr_as_ref(dag) };
+    // SAFETY: Per the documentation the qubits and params pointers are arrays of num_qubits()
+    // and num_params() elements respectively.
+    unsafe {
+        let qargs: &[Qubit] = match gate.num_qubits() {
+            0 => &[],
+            1 => &[Qubit(*qubits.wrapping_add(0))],
+            2 => &[
+                Qubit(*qubits.wrapping_add(0)),
+                Qubit(*qubits.wrapping_add(1)),
+            ],
+            3 => &[
+                Qubit(*qubits.wrapping_add(0)),
+                Qubit(*qubits.wrapping_add(1)),
+                Qubit(*qubits.wrapping_add(2)),
+            ],
+            4 => &[
+                Qubit(*qubits.wrapping_add(0)),
+                Qubit(*qubits.wrapping_add(1)),
+                Qubit(*qubits.wrapping_add(2)),
+                Qubit(*qubits.wrapping_add(3)),
+            ],
+            // There are no ``QkGate``s > 4 qubits
+            _ => panic!(),
```

Comment:
Should there be a message here? And also do we want this to be recoverable by having an `out_index` parameter and just return `ErrorCode`s?

Reply:
I didn't add a panic message mostly just because this function is mostly just a port of what's in `QkCircuit`, and we don't have a message there.

I think it'd be best to keep things as is regarding error handling, and if we decide it's important to revamp, do it all at once for `QkCircuit` and `QkDag`.